### PR TITLE
add linux disk temperatures via smart.go / ghw - prototype - #182

### PR DESCRIPTION
**Pull request by rare-magma**
This PR addresses the feature request #182 . ~~I'd say it's more of a prototype than anything else, just wanted to show that it is possible.~~

Using [smart.go](https://github.com/anatol/smart.go) makes it simple to retrieve disk temperatures from SMART info in the system.
[ghw](https://github.com/jaypipes/ghw) is used for retrieving the disk model name.

There's at least a couple problems with this approach.
1. smart.go needs sudo permissions to retrieve data.
2. ~~smart.go doesn't provide RAW_VALUE numbers, which is the easier way to get a disk temperature without having to perform some calculations (x - raw_value = value) [more info about this here.](https://www.truenas.com/community/threads/smart-attribute-temperature-value-decoding.21622/post-238333)~~ implemented by smart.go on https://github.com/anatol/smart.go/commit/64dea241c3bb4f952c14af0db5d0ff6a1d9bb36f

Please let me know what you think, thanks!

### CODE REVIEW

File: devices/temp_nix.go [hunk 1]
```diff
@@ -28,6 +33,38 @@ func getTemps(temps map[string]int) map[string]error {
...
+	block, err := ghw.Block()
+	if err != nil {
+		log.Print("error getting block device info")
+		return nil
+	}
+
+	var sata = regexp.MustCompile(`sd.?`)
+	var nvme = regexp.MustCompile(`nvme*`)
+	for _, disk := range block.Disks {
+		switch t := disk.Name; {
```
Comment (rare-magma) [inline devices/temp_nix.go:46]: Haven't found a better way to create a switch depending on the disk type (sda / nvme)...

File: devices/temp_nix.go [hunk 2]
```diff
@@ -28,6 +33,38 @@ func getTemps(temps map[string]int) map[string]error {
 			temps[label] = int(sensor.Temperature)
 		}
 	}
+
+	block, err := ghw.Block()
+	if err != nil {
+		log.Print("error getting block device info")
```
Comment (rare-magma) [inline devices/temp_nix.go:39]: Wasn't sure what is the correct way to log errors

File: devices/temp_nix.go (hunk 1 above)
Comment (anatol) [reply devices/temp_nix.go:46]: A better and more idiomatic way to do it is following code:

```go
for _, disk := range block.Disks {
        fmt.Println(disk.Name)
        dev, err := smart.Open("/dev/" + disk.Name)
        if err != nil {
                fmt.Println(err)
                continue
        }

        switch sm := dev.(type) {
        case *smart.SataDevice:
                _, _ = sm.Identify()
        case *smart.ScsiDevice:
                _, _ = sm.Capacity()
        case *smart.NVMeDevice:
                _, _ = sm.ReadSMART()
        }
}
```

**Comment by anatol** [comment]
> smart.go needs sudo permissions to retrieve data.

accessing a device's IOCTL requires admin priveleges. It is true for other tools like `smartctl`.

If you need to test your tool then using a virtual machine (like QEMU) and running the tests as `root` would help here. See https://github.com/anatol/vmtest library that makes such things easier. Here is how the library is used by smart.go https://github.com/anatol/smart.go/blob/main/integration_test.go

File: devices/temp_nix.go (hunk 1 above)
Comment (rare-magma) [reply devices/temp_nix.go]: thanks for the suggestion @anatol it's definitely a more elegant way to do this.
refactored in 98c69b7dacf85b8b25be909b33c333491ae481ad

File: devices/temp_nix.go [hunk 3]
```diff
@@ -28,6 +32,35 @@ func getTemps(temps map[string]int) map[string]error {
...
+			log.Println(err)
+			continue
+		}
+		defer dev.Close()
+
+		switch sm := dev.(type) {
+		case *smart.SataDevice:
+			data, _ := sm.ReadSMARTData()
+			for _, attr := range data.Attrs {
+				if attr.Id == 194 {
```
Comment (anatol) [inline devices/temp_nix.go:54]: with the latest changes in `ataattr` branch the `data.Attrs` becomes a map and you can make this part a bit nicer as

```go
if attr, ok := range data.Attrs[194]; ok {
    // process the temperature value
    temps[disk.Name+"_"+disk.Model] = int(attr.Value)
}
```

File: devices/temp_nix.go [hunk 4]
```diff
@@ -28,6 +32,35 @@ func getTemps(temps map[string]int) map[string]error {
...
+		dev, err := smart.Open("/dev/" + disk.Name)
+		if err != nil {
+			log.Println(err)
+			continue
+		}
+		defer dev.Close()
+
+		switch sm := dev.(type) {
+		case *smart.SataDevice:
+			data, _ := sm.ReadSMARTData()
```
Comment (anatol) [inline devices/temp_nix.go:52]: do not forget to handle error here

File: devices/temp_nix.go (hunk 3 above)
Comment (rare-magma) [reply devices/temp_nix.go]: this is definitely more elegant! refactored in [3c148a4](https://github.com/xxxserxxx/gotop/pull/221/commits/3c148a4d97d0f54f1d6f464ddc64b03fda1ad5fa)

File: devices/temp_nix.go (hunk 4 above)
Comment (rare-magma) [reply devices/temp_nix.go:51]: good catch, I forgot to add checks after the ReadSmart() calls. Fixed in [3c148a4](https://github.com/xxxserxxx/gotop/pull/221/commits/3c148a4d97d0f54f1d6f464ddc64b03fda1ad5fa)

File: devices/temp_nix.go [hunk 5]
```diff
@@ -42,21 +42,26 @@ func getTemps(temps map[string]int) map[string]error {
...
-			for _, attr := range data.Attrs {
-				if attr.Id == 194 {
-					temps[disk.Name+"_"+disk.Model] = int(attr.Value)
-				}
+			if err != nil {
+				log.Print("error getting smart data")
+				return nil
+			}
+			if attr, ok := data.Attrs[194]; ok {
+				temps[disk.Name+"_"+disk.Model] = int(attr.Value)
```
Comment (anatol) [inline devices/temp_nix.go:57]: And here you might want to use `ParseAsTemperature()` or `ParseAsTemperatureRange()` here.

It is not clear to me how to handle situations when different drives format temperature in a different way. The temperature API might be adjusted.

File: devices/temp_nix.go (hunk 4 above)
Comment (anatol) [reply devices/temp_nix.go:51]: it still does not use `err` variable. The line above must be `data, err := sm.ReadSMARTData()`

Comment (rare-magma) [reply devices/temp_nix.go]: oops! thanks for pointing it out! fixed in b89c1dd748b8ee0913f99cdb16079e0515ae7aba

File: devices/temp_nix.go (hunk 5 above)
Comment (rare-magma) [reply devices/temp_nix.go]: @anatol those new methods are quite handy, making use of them on b89c1dd748b8ee0913f99cdb16079e0515ae7aba

File: devices/temp_nix.go [hunk 6]
```diff
@@ -28,6 +32,45 @@ func getTemps(temps map[string]int) map[string]error {
...
 	}
+
+	block, err := ghw.Block()
+	if err != nil {
+		log.Print("error getting block device info")
+		return nil
+	}
+
+	for _, disk := range block.Disks {
+		dev, err := smart.Open("/dev/" + disk.Name)
```
Comment (anatol) [inline devices/temp_nix.go:43]: btw there is an opportunity for optimization here. You can open the devices only once, at the application start and keep it as an array of smart.dev objects. When you need the temperature you just iterate over the list of devices and get the required temperature.

Comment (rare-magma) [reply devices/temp_nix.go]: @anatol something like this? [e96355d](https://github.com/xxxserxxx/gotop/pull/221/commits/e96355d0b82f4075abe059d94c99ce7eaa6a7946)

File: devices/temp_nix.go [hunk 7]
```diff
@@ -8,13 +8,46 @@ import (
...
+		return err
+	}
+	for _, disk := range block.Disks {
+		dev, err := smart.Open("/dev/" + disk.Name)
+		if err == nil {
+			disks = append(disks, *disk)
+			smDevices = append(smDevices, dev)
+		}
+	}
+	return err
```
Comment (anatol) [inline devices/temp_nix.go:39]: `err` variable here is always `nil` (due to loop `err` variable is shadowed). Just return `nil` as an indication that the function was completed successfully.

File: devices/temp_nix.go [hunk 8]
```diff
@@ -4,13 +4,50 @@
...
+}
+
+func endBlock() error {
+	for _, dev := range smDevices {
+		err := dev.Close()
+		if err != nil {
+			log.Print("error closing device")
+			return err
```
Comment (anatol) [inline devices/temp_nix.go:47]: i do not think it makes sense to return here. The method must close all devices (except those that failed).

**Comment by rare-magma** [comment]
@anatol thanks for your suggestions and advice, I've addressed them in [d7232aa](https://github.com/xxxserxxx/gotop/pull/221/commits/d7232aacb737c6eb0caab50b897c43041c85e5d7)

**Comment by anatol** [comment]
The code looks good now. Thank you!

PS You might want to squash all the changes into a single commit.

**Comment by xxxserxxx** [comment]
This all looks great, thanks for the effort! I'll pull and review the changes this weekend; it's time for another patch release, in any case, and I'll try to include this by next week.

Thanks again

**Comment by rare-magma** [comment]
thanks @xxxserxxx!
I think the updates in go.mod broke something in the Processes table:
image

Otherwise I've been running this branch for a while and seems to work as expected, no issues found anymore.

File: devices/temp_nix.go (hunk 2 above)
Comment (xxxserxxx) [reply devices/temp_nix.go]: Yes, sufficiently for a PR. I try to keep messages translat(able), but adding translated error messages is a non-trivial amount of extra effort, so it's fine.

File: devices/temp_nix.go [hunk 9]
```diff
@@ -28,6 +65,33 @@ func getTemps(temps map[string]int) map[string]error {
...
+				}
+				temps[name] = val
+			}
+		case *smart.NVMeDevice:
+			data, err := sm.ReadSMART()
+			if err != nil {
+				log.Printf("error getting smart data for %s: %s", name, err)
+				continue
+			}
+			temps[name] = int(data.Temperature)
```
Comment (anatol) [inline devices/temp_nix.go:91]: NVME reports the temperature in Kelvins. I believe this line needs to be adjusted.

**Comment by anatol** [comment]
I would like to test this feature with nvme disks. Where is the disk temperature reported? I do not see any disks in "Temperatures" section.

**Comment by rare-magma** [comment]
@anatol the disk temperature should be reported in the "Temperatures" widget. If nothing shows up there then perhaps there's a problem with the implementation for nvme disks. 
Did you notice any errors in the log file? Should be in ~/.cache/gotop/errors.log
Did you run gotop with sudo? smartctl needs root privileges to gather info

**Comment by anatol** [comment]
I was able to get the temperature widget with current `master` branch.

nvme devices temperature was off. here is a patch to fix it https://github.com/xxxserxxx/gotop/pull/225

**Pull request merged**